WordPress/gutenberg
# List View: Displace list view items when dragging (a bit more WYSIWYG)

## What?


Partly addresses: https://github.com/WordPress/gutenberg/issues/56539, follows discussion on https://github.com/WordPress/gutenberg/pull/56579.

> [!NOTE]
> 🚧 This PR is still a bit of a hack — for the moment, let's focus on the usability issues, we can tidy the code up later.

This PR explores one potential approach for displacing list view items when dragging within the list view. Or, to put it another way, while dragging blocks in the list view, move the other blocks out of the way.

## Why?


The goal is to create a more WYSIWYG experience while dragging blocks in the list view. By displacing items, and using a drop indicator that effectively previews what the list item will look like when a user finished dropping their item, the list view should feel a bit more "solid" to use.

## How?


While dragging over the list view, move existing list view items up or down to make room for a drop indicator that looks like what the block looks like once it's dropped. I.e. a drop indicator that is a bit like a preview. Details "how" follows:

* Generate an object keyed by `clientId` containing the global index for all blocks within the list view.
* In the list view branch, for each block, look up the index of the current block and compare it to the index for the block being dragged over. Move blocks out of the way by applying either an `is-above` or `is-below` classname.
* Use `transform: translateY()` rules to offset the blocks so that they move out of the way. By using CSS for this, we get nice and smooth animations which disguises the fact that the dragging behaviour only _really_ updates every `200ms`. This hopefully gives the appearance of smoother drag and drop, without (again hopefully) degrading performance.
* Hide the normal list view drop indicator line, and use a drop indicator that previews the list view item in its final position (make it more WYSIWYG)
* Auto-expand list view items on hover
* Make the drag chip used in the list view be slightly transparent so that you can see where you dropping underneath it



## To-do

* [x] Leave a block sized space when dragging starts (don't move blocks around until the user drags over another block)
* [x] Smoothly handle dealing with the space when dragging outside of the list view area (i.e. don't leave the space where it is)
* [x] Correctly animate when releasing the mouse cursor so that the dragged item smoothly goes back into position (or at least remove the wrong animation that is there at the moment)
* [x] Smoothly deal with the space left when dragging multiple blocks (i.e. can we use a CSS variable instead of hard-coded height for the above / below positioning)
* [x] Come up with a better solution for dragging into collapsed blocks
* [x] Update styling of the drag chip
* [x] Look into performance issue when dragging over editor canvas
* [x] Pick a different level of opacity / make it easier to see where you're dragging
* [x] Come up with a solution for dragging to the desired hierarchy at the end of a container block
* [x] ~Smooth dropping of list view items doesn't appear to be working in the site editor~ Known issue, unrelated to this PR: https://github.com/WordPress/gutenberg/pull/57133/files#r1444129276
* [x] Fix jumpiness in Safari due to scroll-while-dragging behaviour

## Testing Instructions





In a post, page, or template with many blocks, try dragging up and down the list view. How does it feel to use? Also, try dragging and dropping on collapsed blocks — while hovering, the block should expand. Where do you get stuck in using this?

Try a post, page, or template with heavy nesting. How does it feel to drag to the end of a nested block, and adjust the hierarchy of where you're dropping?

## Screenshots or screencast 

<img width="392" alt="image" src="https://github.com/WordPress/gutenberg/assets/14988353/18e99d9e-0468-43f4-8e0e-1d87a43b0e29">

https://github.com/WordPress/gutenberg/assets/14988353/c102a48c-cca3-479d-939e-f8e8812a610b
